kernel/kernel-api/src/main/java/io/delta/kernel/internal/IcebergCompatV2Utils.java

```diff
@@ -33,4 +66,216 @@ public static void validateDataFileStatus(DataFileStatus dataFileStatus) {
+          if (Boolean.valueOf(
+              newMetadata
+                  .getConfiguration()
+                  .getOrDefault("delta.enableIcebergCompatV1", "false"))) {
+            throw new IllegalArgumentException(
+                "Not expected to have delta.enableIcebergCompatV1 enabled");
+          }
+        }
+      };
+
+  private static final IcebergCompatCheck ICEBERG_COMPAT_V2_CHECK_HAS_SUPPORTED_TYPES =
+      new IcebergCompatCheck() {
+        @Override
+        void check(boolean isCreatingNewTable, Metadata newMetadata, Protocol newProtocol) {
+          // TODO
+        }
+      };
+
+  private static final IcebergCompatCheck ICEBERG_COMPAT_V2_CHECK_HAS_ALLOWED_PARTITION_TYPES =
+      new IcebergCompatCheck() {
+        @Override
+        void check(boolean isCreatingNewTable, Metadata newMetadata, Protocol newProtocol) {
+          // TODO
+        }
+      };
+
+  private static final IcebergCompatCheck ICEBERG_COMPAT_V2_CHECK_HAS_NO_PARTITION_EVOLUTION =
+      new IcebergCompatCheck() {
+        @Override
+        void check(boolean isCreatingNewTable, Metadata newMetadata, Protocol newProtocol) {
+          // TODO
+        }
+      };
+
+  private static final IcebergCompatCheck ICEBERG_COMPAT_V2_CHECK_HAS_NO_DELETION_VECTORS =
+      new IcebergCompatCheck() {
+        @Override
+        void check(boolean isCreatingNewTable, Metadata newMetadata, Protocol newProtocol) {
+          // TODO
+        }
+      };
+
+  private static class IcebergCompatMetadataValidatorAndUpdater {
+    private final TableConfig<Boolean> requiredDeltaTableProperty;
+    private final List<TableFeature> requiredDependencyTableFeatures;
+    private final List<RequiredDeltaTablePropertyEnforcer> requiredDeltaTableProperties;
+    private final List<IcebergCompatCheck> icebergCompatChecks;
+
+    IcebergCompatMetadataValidatorAndUpdater(
+        TableConfig<Boolean> requiredDeltaTableProperty,
+        List<TableFeature> requiredDependencyTableFeatures,
+        List<RequiredDeltaTablePropertyEnforcer> requiredDeltaTableProperties,
+        List<IcebergCompatCheck> icebergCompatChecks) {
+      this.requiredDeltaTableProperty = requiredDeltaTableProperty;
+      this.requiredDependencyTableFeatures = requiredDependencyTableFeatures;
+      this.requiredDeltaTableProperties = requiredDeltaTableProperties;
+      this.icebergCompatChecks = icebergCompatChecks;
+    }
+
+    boolean isEnabled(Metadata metadata) {
+      return requiredDeltaTableProperty.fromMetadata(metadata);
+    }
+
+    Optional<Metadata> validateAndUpdateMetadata(
+        boolean isCreatingNewTable, Metadata newMetadata, Protocol newProtocol) {
+      Optional<Metadata> updatedMetadata = Optional.empty();
+
+      // table property checks and metadata updates
+      for (RequiredDeltaTablePropertyEnforcer requiredDeltaTableProperty :
+          requiredDeltaTableProperties) {
```

Comment:
Since these refer to independent properties there's never going to be dependencies between the two right?

Reply:
Yes, so far thats the assumption.